sync-core/src/main/java/com/cloudant/sync/query/UnindexedMatcher.java

```diff
@@ -98,14 +106,221 @@ public static UnindexedMatcher matcherWithSelector(Map<String, Object> selector)
+            // to invert the match status of every candidate, rather than
+            // just flipping the result at the end.
+            //
+            // This is because { "$not": { "$eq": "white_cat" } } needs
+            // to be taken as an atomic check, meaning:
+            //   "there's an item in the array that matches `!= "white_cat"`"
+            // rather than:
+            //   "not (there's an item that matches white_cat)"
+            // The latter is satisfied using the $nin operator.
+            if (actual instanceof List) {
+                boolean currentItemPassed = false;
+                for (Object item: (List<Object>) actual) {
+                    // OR as any value in the array can match
+                    currentItemPassed = valueCompare(item, operator, expected);
+                    passed = passed || (invertResult ? !currentItemPassed : currentItemPassed);
+                }
+            } else {
+                passed = valueCompare(actual, operator, expected);
+                passed = invertResult ? !passed : passed;
+            }
+
+            return passed;
+        } else {
+            // We constructed the tree, so shouldn't end up here; error if we do.
+            String msg = String.format("Found unexpected selector execution tree: %s", node);
+            logger.log(Level.SEVERE, msg);
+            return false;
+        }
+    }
+
+    private boolean valueCompare(Object actual, String operator, Object expected) {
+        boolean passed;
+
+        if (operator.equals("$eq")) {
+            passed = compareEq(actual, expected);
+        } else if (operator.equals("$ne")) {
+            passed = compareNE(actual, expected);
+        } else if (operator.equals("$lt")) {
+            passed = compareLT(actual, expected);
+        } else if (operator.equals("$lte")) {
+            passed = compareLTE(actual, expected);
+        } else if (operator.equals("$gt")) {
+            passed = compareGT(actual, expected);
+        } else if (operator.equals("$gte")) {
+            passed = compareGTE(actual, expected);
+        } else if (operator.equals("$exists")) {
+            boolean expectedBool = (Boolean) expected;
+            boolean exists = (actual != null);
+            passed = (exists == expectedBool);
+        } else {
+            String msg = String.format("Found unexpected operator in selector: %s", operator);
+            logger.log(Level.WARNING, msg);
+            passed = false;
+        }
+
+        return passed;
+    }
+
+    private boolean compareEq(Object l, Object r) {
+        if (l instanceof String && r instanceof String) {
+            return l.equals(r);
+        }
+        if (l instanceof Boolean && r instanceof Boolean) {
+            return l == r;
+        }
+        return l instanceof Number && r instanceof Number &&
+                !(l instanceof Float || r instanceof Float) &&
+                ((Number) l).longValue() == ((Number) r).longValue();
```

Comment:
Seems like this will compare `1.0` and `1.1` as equal. I think this needs making protected and some tests written.